iso15118/shared/messages/iso15118_20/common_types.py

```diff
@@ -174,6 +174,125 @@ class RationalNumber(BaseModel):
+
+        Note:
+            In ISO 15118-2, the min limit is 0 and the max is
+            dependent on the field.
+            In ISO 15118-20, the limits are defined by the
+            RationalNumber type which are the limits of the
+            int16 range: [-32768, 32767].
+            Also, in ISO 15118-20, the exponent assumes the range
+            of [-128, 127], but currently that is not being
+            taken into consideration.
+        Args:
+            value (float): parameter value
+
+        Returns:
+           a tuple where the first item is the exponent and other is the
+           int value with the applied exponent
+        """
+        if float_value < 0 and min_limit == 0:
+            raise ValueError(
+                "The conversion of negative values "
+                "for a range with a min limit of 0, is invalid!"
+            )
+        # This is to ensure that we keep the value to be converted, within
+        # the range of the classes RationalNumber and PhysicalValue 'value' data type
+        # int16
+        min_limit = max(min_limit, INT_16_MIN)
+        # For example, the class PVEVSEMaxPowerLimit has a _max_limit = 200000,
+        # so, if the value to be converted is 200000, we cant represent it
+        # with exponent = 0; instead, we need to take into account
+        # the INT_16_MAX = 32768 and represent the value as
+        # value = 20000, exponent = 1
+        max_limit = min(max_limit, INT_16_MAX)
+
+        exponent = 0
+        limit = max_limit
+
+        if float_value < 0:
+            limit = abs(min_limit)
+
+        abs_value = abs(float_value)
+
+        if 0.1 < (abs_value / limit) <= 1 or abs_value == 0.0:
+            # If the first clause is validated, it means that the value
+            # and the max limit/range of the data type have the same
+            # magnitude/exponent and the value is lower than the limit.
+            # eg the value 234 and the limit 500
+            # have the same magnitude = 10**2
+            # 234/500 = 0.468, which means 234 <= 500. Thus, no optimization
+            # is needed, we can return the given value and exponent = 0
+            return exponent, int(float_value)
+
+        # From here on, we try to approximate as much as possible the
+        # value to the max limit, so that we can mitigate as much as
+        # possible the loss of representation of the number
+        if abs_value < limit:
+            while abs_value < limit and exponent > -3:
+                exponent -= 1
+                abs_value = abs_value * 10
+            if abs_value > limit:
+                # for example for a value of 2.34 and max limit = 400
+                # with exponent = -3 we would send an end value of 2340,
+                # which does not fit the limit, so we go back one exponent
+                # to be able to satisfy the type limitation:
+                # exponent = -2 and end value = 234
```

Comment:
I think I found a better algorithm:

```python
import math
def float_to_int_and_exponent(
    value: float,
    min_limit: float = INT_16_MIN,
    max_limit: float = INT_16_MAX
) -> Tuple[int, int]:

    # if the value is 0 return immediately 
    if value == 0:
        return 0, 0

    # Check the sign of the value to determine how to round the result
    round_func = math.floor if value > 0 else math.ceil

    # Iterate over possible exponents to find the best approximation
    for exponent in range(-3, 4):
        new_value = round_func(value * 10 ** (-exponent))
        # If new_value after rounding fits within the range of a 16-bit integer, return it
        if min_limit <= new_value <= max_limit:
            return new_value, exponent

    # If no exponent could make the value fit in a 16-bit integer range,
    # return the original value with exponent 0
    return round_func(value), 0
```

Reply:
Ok, replaced the method with this now. I moved the test for this over from josev_pro land and it passed.